src/gc.c

```diff
@@ -3286,6 +3305,49 @@ static int _jl_gc_collect(jl_ptls_t ptls, jl_gc_collection_t collection)
         gc_num.last_incremental_sweep = gc_end_time;
     }
 
+    int thrashing = 0; // maybe we should report this to the user or error out?
+    size_t heap_size = jl_atomic_load_relaxed(&gc_heap_stats.heap_size);
+    double target_allocs = 0.0;
+    double min_interval = default_collect_interval;
+    if (collection == JL_GC_AUTO) {
+        uint64_t alloc_diff = before_free_heap_size - old_heap_size;
+        uint64_t freed_diff = before_free_heap_size - heap_size;
+        double alloc_smooth_factor = 0.95;
+        double collect_smooth_factor = 0.5;
+        double tuning_factor = 0.03;
+        double alloc_mem = jl_gc_smooth(old_alloc_diff, alloc_diff, alloc_smooth_factor);
+        double alloc_time = jl_gc_smooth(old_mut_time, mutator_time, alloc_smooth_factor);
+        double gc_mem = jl_gc_smooth(old_freed_diff, freed_diff, collect_smooth_factor);
+        double gc_time = jl_gc_smooth(old_pause_time, pause, collect_smooth_factor);
+        old_alloc_diff = alloc_diff;
+        old_mut_time = mutator_time;
+        old_freed_diff = freed_diff;
+        old_pause_time = pause;
+        old_heap_size = heap_size;
+        thrashing = gc_time > mutator_time * 98 ? 1 : 0;
+        if (alloc_mem != 0 && alloc_time != 0 && gc_mem != 0 && gc_time != 0 ) {
+            double alloc_rate = alloc_mem/alloc_time;
+            double gc_rate = gc_mem/gc_time;
+            target_allocs = sqrt(((double)heap_size/min_interval * alloc_rate)/(gc_rate * tuning_factor)); // work on multiples of min interval
+        }
+    }
+    if (target_allocs == 0.0 || thrashing) // If we are thrashing go back to default
+        target_allocs = 2*sqrt((double)heap_size/min_interval);
+
+    uint64_t target_heap = (uint64_t)target_allocs*min_interval + heap_size;
```

Comment:
this will be equivalent to sqrt(heap_size*alloc_rate/(gc_rate*tuning_factor)) * sqrt(min_interval) + heap_size. 
I am a bit confused - shouldnt this be independent of min_interval?

Reply:
I used it as a scaling factor basically. It made it easier for me to reason about what taking the sqrt meant.